lib/spack/docs/packaging_guide.rst

```diff
@@ -373,6 +373,102 @@ some examples:
 In general, you won't have to remember this naming convention because
 :ref:`spack-create` and :ref:`spack-edit` handle the details for you.
 
+-----------------
+Trusted Downloads
+-----------------
+
+Spack verifies that the source code it downloads is not corrupted or
+compromised; or at least, that it is the same version the author of
+the Spack package saw when the package was created.  If Spack uses a
+download method it can verify, we say the download method is
+*trusted*.  Trust is important for *all downloads*: Spack
+has no control over the security of the various sites from which it
+downloads source code, and can never assume that any particular site
+hasn't been compromised.
+
+Trust is established in different ways for different download method.
+For the most common download method --- a single-file tarball --- the
+tarball is checsummed.  Git downloads using ``commit=`` are trusted
+implicitly, as long as a hash is specified.
+
+Spack also provides untrusted download methods: tarball URLs may be
+supplied without a checksum, or Git downloads my specify a branch or
+tag instead of a hash.  If the user does not control or trust the
+source an untrusted download, it is a security risk.  Unless otherwise
+specified by the user for special cases, Spack should by default use
+*only* trusted download methods.
+
+Unfortunately, Spack does not currently provide that guarantee.  It
+does provide the following mechanisms for safety:
+
+#. By default, Spack will only install a tarball package if it has a
+   checksum and that checksum matches.  You can override this with
+   ``spack install --no-checksum``.
+
+#. Numeric versions are almost always tarball downloads, whereas
+   non-numeric versions not named ``develop`` frequently download
+   untrusted branches or tags from a version control system.  As long
+   as a package has at least one numeric version, and no non-numeric
+   version named ``develop``, Spack will prefer it over any
+   non-numeric versions.
```

Comment:
The problem here is, the current code is buggy in figuring out which
version to use.  It cannot stay as-is, since it breaks many attempts to set
a particular version in `packages.py`.  #1561 fixes this.  It, or some
modified version of it, NEEDS to be merged.  (Ping @tgamblin).  I've
documented post-#1561 because this needs to get merged, and it meets all
requirements and concerns brought up in the discussion.

The core problem is that non-specific understandings like this do not
translate into an algorithm that works consistently and predictably in all
cases.  You can't both prefer and not prefer something during
concretization:

The only difference was that Spack was smart enough to understand that
> "develop" was newer, but should not be preferred during concretization.
> @davydden <https://github.com/davydden> Is this right?


There are two layers of ordering of versions: in core comparisons, and then
in the code that chooses a version.  #1561 involves changes to both of
those sections of code.

For core comparisons, the idea is:
    develop > numeric > other non-numeric

For concretization, the comparison is a bit more complex; a version is to
be preferred not just by its ordering, but also by WHERE it came from.  For
example, versions specified in `packages.py` become "greater" in that
ordering than versions that just showed up in a package.  I think the code
makes it pretty clear what's going on there.

As things stand now, a ``develop`` version WILL be chosen if it's in the
``package.py`` and the user doesn't specify anything else.  I think this is
the best we can/should do for now in a consistent manner, until we upgrade
Spack's security; see #1696.  Once that's done, the ``develop`` version
will no longer accidentally be used by casual users because it will almost
certainly not be trusted.  Until then, I think it's best we don't submit
packages with ``develop`` version in them.

-- Elizabeth



On Thu, Sep 8, 2016 at 2:17 PM, Denis Davydov <notifications@github.com>
wrote:

> In lib/spack/docs/packaging_guide.rst:
>
> > +specified by the user for special cases, Spack should by default use
> > +*only* trusted download methods.
> > +
> > +Unfortunately, Spack does not currently provide that guarantee.  It
> > +does provide the following mechanisms for safety:
> > +
> > +#. By default, Spack will only install a tarball package if it has a
> > +   checksum and that checksum matches.  You can override this with
> > +   ``spack install --no-checksum``.
> > +
> > +#. Numeric versions are almost always tarball downloads, whereas
> > +   non-numeric versions not named ``develop`` frequently download
> > +   untrusted branches or tags from a version control system.  As long
> > +   as a package has at least one numeric version, and no non-numeric
> > +   version named ``develop``, Spack will prefer it over any
> > +   non-numeric versions.
>
> The only difference was that Spack was smart enough to understand that
> "develop" was newer, but should not be preferred during concretization.
> @davydden <https://github.com/davydden> Is this right?
>
> to the best of my knowledge, this is the current behaviour.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/1676/files/94896e9e3d3aff3e1f727ee895d333f3ff46ec6d#r78060795>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd5QnJ3YGpkOkCV3MD5eATJhVwHqQks5qoFFFgaJpZM4JxOaa>
> .
>